dev/generate-cuda.py

```diff
@@ -170,23 +169,23 @@ def traverse(node, args={}, forvars=[], declared=[]):  # noqa: B006
             left = getthread_dim(forvars.index(left))
         if right in forvars:
             right = getthread_dim(forvars.index(right))
-        code = "({0} {1} {2})".format(
+        code = "({} {} {})".format(
             left,
             traverse(node.op, args, copy.copy(forvars), declared),
             right,
         )
     elif node.__class__.__name__ == "UnaryOp":
         if node.op.__class__.__name__ == "USub":
-            code = "-{0}".format(
+            code = "-{}".format(
                 traverse(node.operand, args, copy.copy(forvars), declared)
             )
         elif node.op.__class__.__name__ == "Not":
-            code = "!{0}".format(
+            code = "!{}".format(
                 traverse(node.operand, args, copy.copy(forvars), declared)
             )
         else:
             raise Exception(
-                "Unhandled UnaryOp node {0}. Please inform the developers.".format(
+                "Unhandled UnaryOp node {}. Please inform the developers.".format(
                     node.op.__class__.__name__
                 )
```

Comment:
This would have been just fine as an f-string, but pyupgrade is conservative about conversions.

Reply:
Also, Exceptions should always be more specific - if not using a custom one, at least RuntimeError. :)

I also try to always use `msg = <stuff>`, then `raise Exception(msg)`, otherwise your message gets repeated twice, once in the line print and once in the message printout.

Reply:
I was going to say that I only ever use "Exception" in debugging (just like "print", it's easy to search for when cleaning up). This was not written by me. The three exceptions look like they ought to be "NotImplementedError".

However, dev/generate-cuda.py is also on the chopping block for code that will be removed soon. After @swishdiff finalizes the form of the CUDA source code that dev/generate-cuda.py creates, we'll check in that source code and remove the generating script. Henceforth, the source code would be maintained by hand. (This is just seeding many but not all of the initial implementations to save typing for the easy ones.)

Reply:
> it's easy to search for when cleaning up

That's a nice idea. Not worried about changing these, just noting them in passing.